libpod/image/pull.go

```diff
@@ -114,64 +114,113 @@ func (ir *Runtime) getSinglePullRefPairGoal(srcRef types.ImageReference, destNam
+		// Note that we are looking only at the first item of the
+		// refList as Podman normalizes short images differently than
+		// c/image.  While c/image is normalizing to docker.io (just as
+		// Docker), Podman is normalizing to "localhost/".  Hence, if
+		// RepoTags are present we normalize them OR we generate the
+		// digest and use that as a destination name.
+		ref := refList[0]
+		destNames, err := reader.ManifestTagsForReference(ref)
 		if err != nil {
 			return nil, err
 		}
-		defer tarSource.Close()
-		manifest, err := tarSource.LoadTarManifest()
 
-		if err != nil {
-			return nil, errors.Wrapf(err, "error retrieving manifest.json")
-		}
-		// to pull the first image stored in the tar file
-		if len(manifest) == 0 {
-			// use the hex of the digest if no manifest is found
-			reference, err := getImageDigest(ctx, srcRef, sc)
+		if len(destNames) == 0 {
+			destName, err := getImageDigest(ctx, ref, sc)
 			if err != nil {
 				return nil, err
 			}
-			return ir.getSinglePullRefPairGoal(srcRef, reference)
+			destNames = append(destNames, destName)
+		} else {
+			for i := range destNames {
+				ref, err := NormalizedTag(destNames[i])
+				if err != nil {
+					return nil, err
+				}
+				destNames[i] = ref.String()
+			}
 		}
 
-		if len(manifest[0].RepoTags) == 0 {
-			// If the input image has no repotags, we need to feed it a dest anyways
-			digest, err := getImageDigest(ctx, srcRef, sc)
+		for _, destName := range destNames {
+			destRef, err := is.Transport.ParseStoreReference(ir.store, destName)
 			if err != nil {
-				return nil, err
+				return nil, errors.Wrapf(err, "error parsing dest reference name %#v", destName)
 			}
-			return ir.getSinglePullRefPairGoal(srcRef, digest)
+			if srcRef != nil {
+				ref = srcRef
+			}
+			pair := pullRefPair{
+				image:  destName,
+				srcRef: ref,
+				dstRef: destRef,
+			}
+			refPairs = append(refPairs, pair)
 		}
+	}
 
-		// Need to load in all the repo tags from the manifest
-		res := []pullRefPair{}
-		for _, dst := range manifest[0].RepoTags {
-			//check if image exists and gives a warning of untagging
-			localImage, err := ir.NewFromLocal(dst)
-			imageID := strings.TrimSuffix(manifest[0].Config, ".json")
-			if err == nil && imageID != localImage.ID() {
-				logrus.Errorf("the image %s already exists, renaming the old one with ID %s to empty string", dst, localImage.ID())
-			}
+	return refPairs, nil
+}
 
-			pullInfo, err := ir.getPullRefPair(srcRef, dst)
-			if err != nil {
-				return nil, err
+// getPullRefPairFromDockerReference returns a slice of pullRefPairs for the specified docker reference.
+func (ir *Runtime) getPullRefPairsFromDockerReference(ctx context.Context, ref types.ImageReference, sc *types.SystemContext) ([]pullRefPair, error) {
+	// Let's first assume it's a file-only reference.
+	reader, err := archive.NewReader(sc, ref.StringWithinTransport())
+	if err == nil {
+		defer func() {
+			if err := reader.Close(); err != nil {
```

Comment:
This closes the `reader` before any references are used; that won’t work.

Reply:
It's working in this case, I am enforcing the original reference.

What I am trying to do here is to extract all RepoTags from a given reference. I didn't find a way other than using an `archive.Reader`.

Reply:
> It's working in this case, I am enforcing the original reference.

Ah, yes. But that’s a very localized override, so `getPullRefPairsFromDockerArchiveReader` also accepts multi-image archives, and silently creates (that-original-reference, any-tag-for-any-image-in-archive) pairs AFAICS.

> What I am trying to do here is to extract all RepoTags from a given reference.

`NewReaderForReference`+`ManifestTagsForReference`.

> I didn't find a way other than using an `archive.Reader`.

Yes, that’s intentional, to make it annoyingly difficult to extract `RepoTags` without creating the `archive.Reader`, so that the easy way to use the API does not involve creating the temporary archive twice. 

(If you _really don’t want to_ keep a `Reader` alive throughout the pull, `docker archive.New{Index,}Reference` does allow creating reader-independent references for the individual names/images in an archive. But even a _short-lived_ Reader for the purpose of the heuristics creating ref pairs would speed up things like `getImageDigest`, so please use that, and maybe disassociate from the `Reader` only for the ref pair.)

Reply:
Thanks again for the guidance. The reader now persists across the entire pull (moved back the cleanUp idea we had earlier).